src/main/scala/analysis/InterprocSteensgaardAnalysis.scala

```diff
@@ -0,0 +1,395 @@
+   */
+  def exprToRegion(expr: Expr, n: CfgCommandNode): Set[MemoryRegion] = {
+    var res = Set[MemoryRegion]()
+    mmm.popContext()
+    mmm.pushContext(n.parent.data.name)
+    expr match { // TODO: Stack detection here should be done in a better way or just merged with data
+      case binOp: BinaryExpr if binOp.arg1 == stackPointer =>
+        evaluateExpressionWithSSA(binOp.arg2, constantProp(n)).foreach {
+          case b: BitVecLiteral =>
+            if binOp.arg2.variables.exists { case v: Variable => v.sharedVariable } then {
+              println("Shared stack object: " + b)
+              println("Shared in: " + expr)
+              val regions = mmm.findSharedStackObject(b.value)
+              println("found: " + regions)
+              res = res ++ regions
+            } else {
+              val region = mmm.findStackObject(b.value)
+              if (region.isDefined) {
+                res = res + region.get
+              }
+            }
+        }
+        res
+      case binaryExpr: BinaryExpr =>
+        res = res ++ reducibleToRegion(binaryExpr, n)
+        res
+      case _ =>
+        evaluateExpressionWithSSA(expr, constantProp(n)).foreach {
+          case b: BitVecLiteral =>
+            println("BitVecLiteral: " + b)
+            val region = mmm.findDataObject(b.value)
+            if (region.isDefined) {
+              res = res + region.get
+            }
+        }
+        if (res.isEmpty && expr.isInstanceOf[Variable]) { // may be passed as param
+          val ctx = regionAccesses(n)
+          if (ctx.contains(RegisterVariableWrapper(expr.asInstanceOf[Variable]))) {
+            ctx(RegisterVariableWrapper(expr.asInstanceOf[Variable])) match {
+              case FlatEl(al) =>
+                al match
+                  case load: MemoryLoad => // treat as a region
+                    res = res ++ exprToRegion(load.index, n)
+                  case binaryExpr: BinaryExpr =>
+                    res = res ++ reducibleToRegion(binaryExpr, n)
+                    res = res ++ exprToRegion(al, n)
+                  case _ => // also treat as a region (for now) even if just Base + Offset without memLoad
+                    res = res ++ exprToRegion(al, n)
+            }
+          }
+        }
+        res
+    }
+  }
+
+  /** @inheritdoc
+    */
+  def analyze(): Unit =
+    // generate the constraints by traversing the AST and solve them on-the-fly
+    cfg.nodes.foreach(visit(_, ()))
+
+  /** Generates the constraints for the given sub-AST.
+    * @param node
+    *   the node for which it generates the constraints
+    * @param arg
+    *   unused for this visitor
+    */
+  def visit(n: CfgNode, arg: Unit): Unit = {
+
+    def varToStTerm(vari: VariableWrapper): Term[StTerm] = IdentifierVariable(vari)
+    def exprToStTerm(expr: MemoryRegion | Expr): Term[StTerm] = ExpressionVariable(expr)
+    def allocToTerm(alloc: MemoryRegion): Term[StTerm] = AllocVariable(alloc)
```

Comment:
Why do these exist when the constructors for those case classes could be called each time instead?